core/src/main/java/org/opensearch/sql/planner/optimizer/rule/EvalPushDown.java

```diff
@@ -42,6 +51,38 @@ public class EvalPushDown<T extends LogicalPlan> implements Rule<T> {
                 return logicalEval;
               });
 
+  public static final Rule<LogicalSort> PUSH_DOWN_SORT =
+      match(sort(evalCapture()))
+          .apply(
+              (sort, logicalEval) -> {
+                List<LogicalPlan> child = logicalEval.getChild();
+                Map<ReferenceExpression, Expression> evalExpressionMap =
+                    logicalEval.getExpressions().stream()
+                        .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
+                List<Pair<SortOption, Expression>> sortList = sort.getSortList();
+                List<Pair<SortOption, Expression>> newSortList = new ArrayList<>();
+                for (Pair<SortOption, Expression> pair : sortList) {
+                  /*
+                   Narrow down the optimization to only support:
+                   1. The expression in sort and replaced expression are both ReferenceExpression.
+                   2. No internal reference in eval.
+                  */
+                  if (pair.getRight() instanceof ReferenceExpression) {
+                    ReferenceExpression ref = (ReferenceExpression) pair.getRight();
+                    Expression replacedExpr = evalExpressionMap.getOrDefault(ref, ref);
+                    if (replacedExpr instanceof ReferenceExpression) {
+                      ReferenceExpression newRef = (ReferenceExpression) replacedExpr;
+                      if (!evalExpressionMap.containsKey(newRef)) {
+                        newSortList.add(Pair.of(pair.getLeft(), newRef));
+                      } else return sort;
+                    } else return sort;
+                  } else return sort;
```

Comment:
I think you can `return sort` at the end of `for` block once.

Reply:
Then we need a flag to indicate returning sort or not for each iteration. 

It could be simplified by pattern matching with java version>=14. But since we also need to support java11 for 2.x, I don't use that feature to keep code align.